lib/src/message_signing.dart

```diff
@@ -0,0 +1,428 @@
+import 'dart:collection';
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:crypto/crypto.dart';
+
+import 'structured_fields.dart';
+
+/// Signature algorithms supported by the Approov message signing flow.
+enum SignatureAlgorithm {
+  hmacSha256,
+  ecdsaP256Sha256,
+}
+
+SfItem _buildComponentIdentifier(String value, Map<String, dynamic>? parameters) {
+  return SfItem.string(value, parameters);
+}
+
+String _componentIdentifierValue(SfItem item) {
+  final bareItem = item.bareItem;
+  if (bareItem.type != SfBareItemType.string) {
+    throw StateError('Component identifiers must be sf-string values');
+  }
+  return bareItem.value as String;
+}
+
+/// Holds configuration for message signature parameters, mirroring the Swift implementation.
+class SignatureParameters {
+  SignatureParameters()
+      : _componentIdentifiers = <SfItem>[],
+        _parameters = LinkedHashMap<String, SfBareItem>();
+
+  SignatureParameters.copy(SignatureParameters other)
+      : _componentIdentifiers = List<SfItem>.from(other._componentIdentifiers),
+        _parameters = LinkedHashMap<String, SfBareItem>.from(other._parameters),
+        debugMode = other.debugMode,
+        algorithm = other.algorithm;
+
+  final List<SfItem> _componentIdentifiers;
+  final LinkedHashMap<String, SfBareItem> _parameters;
+
+  bool debugMode = false;
+  SignatureAlgorithm algorithm = SignatureAlgorithm.hmacSha256;
+
+  List<SfItem> get componentIdentifiers => List.unmodifiable(_componentIdentifiers);
+
+  void addComponentIdentifier(String identifier, {Map<String, dynamic>? parameters}) {
+    final normalized = identifier.startsWith('@') ? identifier : identifier.toLowerCase();
+    final candidateParameters = SfParameters(parameters);
+    if (_componentIdentifiers.any(
+      (item) =>
+          _componentIdentifierMatches(item, normalized, candidateParameters),
+    )) {
+      return;
+    }
+    _componentIdentifiers.add(_buildComponentIdentifier(normalized, parameters));
+  }
+
+  bool _componentIdentifierMatches(SfItem item, String value, SfParameters candidate) {
+    if (_componentIdentifierValue(item) != value) return false;
+    return _parametersMatch(item.parameters, candidate);
+  }
+
+  bool _parametersMatch(SfParameters existing, SfParameters candidate) {
+    final existingMap = existing.asMap();
+    final candidateMap = candidate.asMap();
+    if (existingMap.length != candidateMap.length) return false;
+    for (final entry in candidateMap.entries) {
+      final existingValue = existingMap[entry.key];
+      if (existingValue == null) return false;
+      if (existingValue.serialize() != entry.value.serialize()) return false;
+    }
+    return true;
+  }
+
+  void setAlg(String value) {
+    _parameters['alg'] = SfBareItem.string(value);
```

Comment:
'alg' and 'algorithm' should be the same value in your implementation. (I think algorithm should probably be removed). The SignatureParameters object shouldn't really know about the set of vaild algorithms either. That should be managed at a higher level (the default Approov message singing implementation - the same approach taken by okhttp). If this has been copied from the swift stuff, then I am a little concerned that we have different implementations betwee android and iOS.